client/redux/cart.js

```diff
@@ -0,0 +1,119 @@
+import axios from 'axios';
+import { ADD_CART_ITEM, SET_CART_ITEMS, ADD_MULTIPLE_CART_ITEMS, CHANGE_CART_ITEM, REMOVE_CART_ITEM } from './actionConstants';
+import { addLocalCartItem, changeLocalCartItem, removeLocalCartItem } from '../localCart/'
+
+// ACTION CREATORS
+export const _addCartItem = cartItem => {
+  return {
+    type: ADD_CART_ITEM,
+    cartItem
+  }
+};
+
+export const _changeCartItem = cartItem => {
+  return {
+    type: CHANGE_CART_ITEM,
+    cartItem
+  }
+};
+
+export const _removeCartItem = cartItem => {
+  return {
+    type: REMOVE_CART_ITEM,
+    cartItem
+  }
+};
+
+export const _setCartItems = cartItems => {
+  return {
+    type: SET_CART_ITEMS,
+    cartItems
+  }
+}
+
+export const _addMultipleCartItems = cartItems => {
+  return {
+    type: ADD_MULTIPLE_CART_ITEMS,
+    cartItems
+  }
+}
+
+// THUNK CREATORS
+export const addCartItem = (cartItem, isLoggedIn = false) => {
+  return async (dispatch) => {
+    let newCartItem
+    //do two different things depending on whether user is loggedIn
+    //if logged in, send to the DB, which will return the new cartItem
+    if(isLoggedIn) newCartItem = (await axios.post('/api/cart/item', cartItem)).data;
+    //if not logged in, write to localStorage
+    else newCartItem = addLocalCartItem(cartItem)
+    //note that the returned cart item will have the correct additional
+    //quantity if the item already in cart
+    dispatch(_addCartItem(newCartItem));
+  }
+};
+
+export const changeCartItem = (cartItem, isLoggedIn = false) => {
+  return async (dispatch) => {
+    let updatedCartItem
+    //do two different things depending on whether user is loggedIn
+    //if logged in, send to the DB, which will return the updated cartItem
+    if(isLoggedIn) updatedCartItem = (await axios.put('/api/cart/item', cartItem)).data;
+    //if not logged in, write to localStorage
+    else updatedCartItem = changeLocalCartItem(cartItem)
+    //note that the returned cart item will have the correct new qty
+    dispatch(_changeCartItem(updatedCartItem));
+  }
+};
+
+export const removeCartItem = (cartItem, isLoggedIn = false) => {
+  return async (dispatch) => {
+    let deleteSuccessful
+    //do two different things depending on whether user is loggedIn
+    //if logged in, send to the DB,
+    //which will **NOT** return the deleted cartItem, so we
+    //have to dispatch the original cart item if we succeed in deleting
+    if(isLoggedIn) deleteSuccessful = (await axios.delete(`/api/cart/item/${cartItem.id}`)).status;
+    //if not logged in, write to localStorage
+    else deleteSuccessful = removeLocalCartItem(cartItem)
+    //if the cart item was deleted, you should get back status 204
+    //then dispatch the original cart item for the reducer
+    if(deleteSuccessful === 204) {
+      dispatch(_removeCartItem(cartItem));
+    }
```

Comment:
i think try catch would be useful here

Reply:
Yeah. Probably all of these should have some kind of "oops, sorry, we can't do that to your cart right now" error handling. Or did you have something else in mind for try/catch?